### PR TITLE
Rewrite unittest-based repository reclaim tests to pytest

### LINKED ISSUE #821: Refactor test_reclaim_disk_space module to use pytest fixtures
The `test_reclaim_disk_space` module[0] needs to be refactored to use pytest fixtures.

[0] https://github.com/pulp/pulp_file/blob/main/pulp_file/tests/functional/api/from_pulpcore/test_reclaim_disk_space.py

### PR DESCRIPTION
**Pull request by lubosmj**
Required PR: https://github.com/pulp/pulpcore/pull/3281

closes #821

### CODE REVIEW

File: pulp_file/tests/functional/api/from_pulpcore/test_reclaim_disk_space.py [hunk 1]
```diff
@@ -1,193 +1,170 @@
...
-        monitor_task(reclaim_response.task)
+        # Publication
+        publication_data = FileFilePublication(repository=repo.pulp_href)
+        publication = gen_object_with_cleanup(file_pub_api_client, publication_data)
 
-        artifacts_after_reclaim = self.artifacts_api.list().count
-        self.assertLess(artifacts_after_reclaim, artifacts_before_reclaim)
+        # Distribution
+        body = gen_distribution()
+        body["publication"] = publication.pulp_href
```
Comment (gerrod3) [inline pulp_file/tests/functional/api/from_pulpcore/test_reclaim_disk_space.py:79]:
```suggestion
        body = gen_distribution(publication=publication.pulp_href)
```

File: pulp_file/tests/functional/api/from_pulpcore/test_reclaim_disk_space.py [hunk 2]
```diff
@@ -1,193 +1,170 @@
...
+    artifacts_before_download = artifacts_api_client.list().count
+    content = get_content(repo.to_dict())[FILE_CONTENT_NAME][0]
+    download_content_unit(distribution.base_path, content["relative_path"])
 
-    def _repo_sync_distribute(self, policy="immediate"):
-        """Helper to create & populate a repository and distribute it."""
-        repo = self.repo_api.create(gen_repo())
-        self.addCleanup(self.repo_api.delete, repo.pulp_href)
+    artifacts = artifacts_api_client.list().count
+    assert artifacts >= artifacts_before_download
```
Comment (gerrod3) [inline pulp_file/tests/functional/api/from_pulpcore/test_reclaim_disk_space.py:105]:
```suggestion
    assert artifacts > artifacts_before_download
```
The original assertion was just greater than.

File: pulp_file/tests/functional/api/from_pulpcore/test_reclaim_disk_space.py [hunk 3]
```diff
@@ -1,193 +1,170 @@
...
+    download_content_unit(distribution.base_path, content[0]["relative_path"])
 
-        # Publication
-        publication_data = FileFilePublication(repository=repo.pulp_href)
-        publication_response = self.publication_api.create(publication_data)
-        task_response = monitor_task(publication_response.task)
-        publication = self.publication_api.read(task_response.created_resources[0])
-        self.addCleanup(self.publication_api.delete, publication.pulp_href)
+    artifacts = artifacts_api_client.list().count
+    assert artifacts >= artifacts_after_reclaim
```
Comment (gerrod3) [inline pulp_file/tests/functional/api/from_pulpcore/test_reclaim_disk_space.py:118]:
```suggestion
    assert artifacts > artifacts_after_reclaim
```

**Review by gerrod3 — CHANGES_REQUESTED** [review]

File: pulp_file/tests/functional/api/from_pulpcore/test_reclaim_disk_space.py [hunk 4]
```diff
@@ -1,193 +1,170 @@
...
+
+from pulp_smash.pulp3.bindings import monitor_task
+from pulp_smash.pulp3.utils import gen_repo, gen_distribution, get_content
+from pulpcore.client.pulp_file import FileFilePublication, RepositorySyncURL
+
 from .constants import FILE_CONTENT_NAME
 
 
-class ReclaimSpaceTestCase(PulpTestCase):
+def test_reclaim_immediate_content(
```
Comment (dkliban) [inline pulp_file/tests/functional/api/from_pulpcore/test_reclaim_disk_space.py:11]: We should mark this test as parallel.

File: pulp_file/tests/functional/api/from_pulpcore/test_reclaim_disk_space.py [hunk 5]
```diff
@@ -1,193 +1,170 @@
...
+    # reclaim disk space
+    reclaim_response = repositories_reclaim_space_api_client.reclaim(
+        {"repo_hrefs": [repo.pulp_href]}
+    )
+    monitor_task(reclaim_response.task)
 
-        artifacts_before_download = self.artifacts_api.list().count
-        content = get_content(repo.to_dict())[FILE_CONTENT_NAME][0]
-        download_content_unit(self.cfg, distribution.to_dict(), content["relative_path"])
+    # assert no artifacts left
```
Comment (dkliban) [inline pulp_file/tests/functional/api/from_pulpcore/test_reclaim_disk_space.py:39]: We should assert that the artifacts that were in the fixture repository are no longer in Pulp. Here is an example[0].

[0] https://github.com/pulp/pulp_file/blob/main/pulp_file/tests/functional/api/test_download_policies.py#L66-L67

File: pulp_file/tests/functional/api/from_pulpcore/test_reclaim_disk_space.py [hunk 6]
```diff
@@ -1,193 +1,170 @@
...
-        self.assertGreater(artifacts, artifacts_before_download)
+    # sync repo again
+    repository_sync_data = RepositorySyncURL(remote=remote.pulp_href)
+    sync_response = file_repo_api_client.sync(repo.pulp_href, repository_sync_data)
+    monitor_task(sync_response.task)
 
-        # reclaim disk space
-        reclaim_response = self.reclaim_api.reclaim({"repo_hrefs": [repo.pulp_href]})
-        monitor_task(reclaim_response.task)
+    # assert re-sync populated missing artifacts
```
Comment (dkliban) [inline pulp_file/tests/functional/api/from_pulpcore/test_reclaim_disk_space.py:48]: We should that the specific artifacts are there. Similar to above.

File: pulp_file/tests/functional/api/from_pulpcore/test_reclaim_disk_space.py [hunk 7]
```diff
@@ -1,193 +1,170 @@
...
-        # Distribution
-        body = gen_distribution()
-        body["publication"] = publication.pulp_href
-        distribution_response = self.distributions_api.create(body)
-        created_resources = monitor_task(distribution_response.task).created_resources
-        distribution = self.distributions_api.read(created_resources[0])
-        self.addCleanup(self.distributions_api.delete, distribution.pulp_href)
 
-        return repo, distribution
+def test_immediate_reclaim_becomes_on_demand(
```
Comment (dkliban) [inline pulp_file/tests/functional/api/from_pulpcore/test_reclaim_disk_space.py:121]: This test can also be marked as parallel.

File: pulp_file/tests/functional/api/from_pulpcore/test_reclaim_disk_space.py [hunk 8]
```diff
@@ -1,193 +1,170 @@
...
+    and then re-populated back after client request.
+    """
+    repo, distribution = sync_repository_distribution(policy="on_demand")
 
-        repos_locked = [r.split(":")[-1] for r in task_status.reserved_resources_record]
-        self.assertEqual(len(repos), len(repos_locked))
-        self.assertEqual(set(repos), set(repos_locked))
+    artifacts_before_download = artifacts_api_client.list().count
+    content = get_content(repo.to_dict())[FILE_CONTENT_NAME][0]
+    download_content_unit(distribution.base_path, content["relative_path"])
```
Comment (dkliban) [inline pulp_file/tests/functional/api/from_pulpcore/test_reclaim_disk_space.py:102]: You should switch from this fixture to the existing method in utils[0].

[0] https://github.com/pulp/pulp_file/blob/main/pulp_file/tests/functional/utils.py#L315-L323

**Review by dkliban — CHANGES_REQUESTED** [review]
I left some comments. I believe most of the tests can be run in parallel. I also suspect that you could possibly combine this workflow with the download policy tests[0]. 

[0] https://github.com/pulp/pulp_file/blob/main/pulp_file/tests/functional/api/test_download_policies.py

**Review by dkliban — APPROVED** [review]

**Review by gerrod3 — APPROVED** [review]

**Pull request merged**